var/spack/repos/builtin/packages/cairo/package.py

```diff
@@ -25,55 +27,187 @@ class Cairo(AutotoolsPackage):
+    # https://github.com/mesonbuild/meson/issues/8224
+    # https://github.com/microsoft/vcpkg/pull/38313
+    depends_on("lzo", when="@1.17.6: build_system=meson")
+
+    with when("@:1.17.6"):
+        depends_on("automake", type="build")
+        depends_on("autoconf", type="build")
+        depends_on("libtool", type="build")
+
+        depends_on("m4", type="build")
+
+        depends_on("pixman@0.36.0:", when="@1.17.2:")
+        depends_on("freetype build_system=autotools", when="+ft")
+        depends_on("fontconfig@2.10.91:", when="+fc")  # Require newer version of fontconfig.
+        depends_on("libpng", when="+png")
+        depends_on("glib")
+        depends_on("which", type="build")
+
+    with when("@1.17.8:"):
+        depends_on("meson@1.3.0:")
+        depends_on("freetype@2.13.0:", when="+ft")
+        depends_on("libpng@1.4.0:", when="+png")
+
+        depends_on("glib@2.14:", when="+gobject")
+        depends_on("pixman@0.40.0:")
+        depends_on("fontconfig@2.13.0:", when="+fc")
+        depends_on("binutils", when="+symbol-lookup")
+
+    with when("+X"):
+        depends_on("libx11")
+        depends_on("libxext")
+
+        depends_on("libxrender")
+        depends_on("libxrender@0.6:", when="@1.17.8:")
+
+        depends_on("libxcb")
+        depends_on("libxcb@1.6:", when="@1.17.8:")
+
+        depends_on("python", type="build")
 
     # patch from https://gitlab.freedesktop.org/cairo/cairo/issues/346
     patch("fontconfig.patch", when="@1.16.0:1.17.2")
     # Don't regenerate docs to avoid a dependency on gtk-doc
-    patch("disable-gtk-docs.patch", when="^autoconf@2.70:")
+    patch("disable-gtk-docs.patch", when="build_system=autotools ^autoconf@2.70:")
+
+
+class MesonBuilder(meson.MesonBuilder):
+    def enable_or_disable(self, feature_name, variant=None):
+        if variant is None:
+            variant = feature_name
+        return (
+            f"-D{feature_name}=enabled"
+            if self.spec.satisfies(f"+{variant}")
+            else f"-D{feature_name}=disabled"
+        )
+
+    def meson_args(self):
+        args = []
+
+        args.append(self.enable_or_disable("dwrite"))
+        args.append(self.enable_or_disable("fontconfig", variant="ft"))
+        args.append(self.enable_or_disable("freetype", variant="fc"))
+
+        # +svg implies png now
+        if self.spec.satisfies("+svg"):
+            self.spec.variants["png"].value = True
+
+        args.append(self.enable_or_disable("png"))
+
+        args.append(self.enable_or_disable("quartz"))
+        args.append(self.enable_or_disable("tee"))
+        args.append(self.enable_or_disable("xcb"))
+
+        args.append(self.enable_or_disable("xlib", variant="X"))
+        args.append(self.enable_or_disable("xlib-xcb", variant="X"))
+
+        # +pdf implies zlib now
+        if self.spec.satisfies("+pdf"):
+            self.spec.variants["zlib"].value = True
```

Comment:
This seems like it could get confusion when stuff is changed under the feet of the users. Can we add a conflict for this instead?
```py
    conficts("~zlib", when="+pdf build_system=meson")
```

Reply:
resolved

Reply:
I've replaced these with requires and refined the conflicts. I think given the sane defaults, the conflicts and requires should keep it mostly in line.